go/fn/run.go

```diff
@@ -28,34 +29,65 @@ func AsMain(p ResourceListProcessor) error {
 	if err != nil {
 		return fmt.Errorf("unable to read from stdin: %v", err)
 	}
-	out, err := Run(p, in)
-	if err != nil {
-		return err
-	}
+	out := Run(p, in)
 	_, err = os.Stdout.Write(out)
 	return err
 }
 
 // Run evaluates the function. input must be a resourceList in yaml format. An
 // updated resourceList will be returned.
-func Run(p ResourceListProcessor, input []byte) ([]byte, error) {
+func Run(p ResourceListProcessor, input []byte) []byte {
 	rl, err := ParseResourceList(input)
 	if err != nil {
-		return nil, err
+		return nil
 	}
 	defer handleOptOrDieErr()
 	err = p.Process(rl)
-	if err != nil {
-		// If the error is not a Results type, we wrap the error as a Result.
-		if results, ok := err.(Results); ok {
-			rl.Results = append(rl.Results, results...)
-		} else if result, ok := err.(Result); ok {
-			rl.Results = append(rl.Results, &result)
-		} else if result, ok := err.(*Result); ok {
-			rl.Results = append(rl.Results, result)
+	defer handleCritical()
+	logResult(rl, err)
+	o, _ := rl.ToYAML()
+	return o
+}
+
+func handleCritical() {
+	if v := recover(); v != nil {
+		if e, ok := v.(Result); ok {
+			log.Fatalf(e.Message)
+		} else if e, ok := v.(error); ok {
+			log.Fatalf(e.Error())
```

Comment:
`_, err = os.Stdout.Write(out)` in `AsMain` function (line 35) writes out the ResourceList to stdout. The results are stored in ResourceList, and writing it out to stdout is how kpt gets the ResourceList. 

The execution in this case will do something like

```
err = p.Process(rl)
logResult(rl, err)
log.Fatal(e.Error()
```

In this case, we are not doing `_, err = os.Stdout.Write(out)` (where out is the ResourceList) anywhere. So kpt will not get the structured Results.

Reply:
+1
We should not exit until we have a chance to write the ResourceList to stdout.

Reply:
Notes from meeting

This approach can work if we do log.Fatal(resourceList), because it will write the resourceList to stdout before exiting.